### PR TITLE
Mazin Vladimir, 2538 group

### PR DESCRIPTION
**Pull request by gelkin**
ver 1.0

### CODE REVIEW

**Comment by thevery** [comment]
no top/bottom padding

**Comment by thevery** [comment]
use 4.x theme for the second screen and remove action/title bar

**Comment by thevery** [comment]
update toast is nice

**Comment by thevery** [comment]
previews are scaled, not cropped

**Comment by thevery** [comment]
UX issues: refresh button takes too much space (should be moved to action bar)

**Comment by thevery** [comment]
app is ok (20-3=17), code review pending

**Comment by thevery** [comment]
E/AndroidRuntime( 8247): java.lang.RuntimeException: Error receiving broadcast Intent { act=com.gmail.mazinva.ACTION flg=0x10 (has extras) } in com.gmail.mazinva.PopularPhotos.MainActivity$1@425a6cc0
E/AndroidRuntime( 8247):    at android.app.LoadedApk$ReceiverDispatcher$Args.run(LoadedApk.java:778)
E/AndroidRuntime( 8247):    at android.os.Handler.handleCallback(Handler.java:733)
E/AndroidRuntime( 8247):    at android.os.Handler.dispatchMessage(Handler.java:95)
E/AndroidRuntime( 8247):    at android.os.Looper.loop(Looper.java:136)
E/AndroidRuntime( 8247):    at android.app.ActivityThread.main(ActivityThread.java:5017)
E/AndroidRuntime( 8247):    at java.lang.reflect.Method.invokeNative(Native Method)
E/AndroidRuntime( 8247):    at java.lang.reflect.Method.invoke(Method.java:515)
E/AndroidRuntime( 8247):    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:779)
E/AndroidRuntime( 8247):    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:595)
E/AndroidRuntime( 8247):    at dalvik.system.NativeStart.main(Native Method)
E/AndroidRuntime( 8247): Caused by: java.lang.IllegalStateException: Couldn't read row 2, col 0 from CursorWindow.  Make sure the Cursor is initialized correctly before accessing data from it.
E/AndroidRuntime( 8247):    at android.database.CursorWindow.nativeGetLong(Native Method)
E/AndroidRuntime( 8247):    at android.database.CursorWindow.getLong(CursorWindow.java:507)
E/AndroidRuntime( 8247):    at android.database.AbstractWindowedCursor.getLong(AbstractWindowedCursor.java:75)
E/AndroidRuntime( 8247):    at android.database.AbstractCursor.moveToPosition(AbstractCursor.java:220)
E/AndroidRuntime( 8247):    at android.database.AbstractCursor.moveToNext(AbstractCursor.java:245)
E/AndroidRuntime( 8247):    at com.gmail.mazinva.PopularPhotos.MainActivity.showImages(MainActivity.java:135)
E/AndroidRuntime( 8247):    at com.gmail.mazinva.PopularPhotos.MainActivity$1.onReceive(MainActivity.java:48)
E/AndroidRuntime( 8247):    at android.app.LoadedApk$ReceiverDispatcher$Args.run(LoadedApk.java:768)
E/AndroidRuntime( 8247):    ... 9 more
W/ActivityManager(  765):   Force finishing activity com.gmail.mazinva.PopularPhotos/.MainActivity

**Comment by gelkin** [comment]
Can I fix my app according to your remarks to get one more point? (so I can
get 60)

app is ok (20-3=17), code review pending

—
Reply to this email directly or view it on
GitHubhttps://github.com/IFMO-MobDev-2013/extratask1/pull/10#issuecomment-32792876
.

**Comment by thevery** [comment]
17 is not final score (it could be higher/lower), because I haven't checked code yet.

**Comment by thevery** [comment]
another crash:

E/AndroidRuntime(25647): Process: com.gmail.mazinva.PopularPhotos, PID: 25647
E/AndroidRuntime(25647): java.lang.RuntimeException: Unable to start activity ComponentInfo{com.gmail.mazinva.PopularPhotos/com.gmail.mazinva.PopularPhotos.SingleImageActivity}: android.database.CursorIndexOutOfBoundsException: Index 0 requested, with a size of 0
E/AndroidRuntime(25647):    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2195)
E/AndroidRuntime(25647):    at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2245)
E/AndroidRuntime(25647):    at android.app.ActivityThread.access$800(ActivityThread.java:135)
E/AndroidRuntime(25647):    at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196)
E/AndroidRuntime(25647):    at android.os.Handler.dispatchMessage(Handler.java:102)
E/AndroidRuntime(25647):    at android.os.Looper.loop(Looper.java:136)
E/AndroidRuntime(25647):    at android.app.ActivityThread.main(ActivityThread.java:5017)
E/AndroidRuntime(25647):    at java.lang.reflect.Method.invokeNative(Native Method)
E/AndroidRuntime(25647):    at java.lang.reflect.Method.invoke(Method.java:515)
E/AndroidRuntime(25647):    at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:779)
E/AndroidRuntime(25647):    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:595)
E/AndroidRuntime(25647):    at dalvik.system.NativeStart.main(Native Method)
E/AndroidRuntime(25647): Caused by: android.database.CursorIndexOutOfBoundsException: Index 0 requested, with a size of 0
E/AndroidRuntime(25647):    at android.database.AbstractCursor.checkPosition(AbstractCursor.java:426)
E/AndroidRuntime(25647):    at android.database.AbstractWindowedCursor.checkPosition(AbstractWindowedCursor.java:136)
E/AndroidRuntime(25647):    at android.database.AbstractWindowedCursor.getBlob(AbstractWindowedCursor.java:44)
E/AndroidRuntime(25647):    at com.gmail.mazinva.PopularPhotos.SingleImageActivity.onCreate(SingleImageActivity.java:27)
E/AndroidRuntime(25647):    at android.app.Activity.performCreate(Activity.java:5231)
E/AndroidRuntime(25647):    at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1087)
E/AndroidRuntime(25647):    at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2159)
E/AndroidRuntime(25647):    ... 11 more

**Comment by gelkin** [comment]
Ok, this error seems to appear after rotating the screen right after
pushing the update-button. So, the only solution that came to my mind is to
create extra database and show one of databases until the other is not
updated correctly. Is it a suitable solution?

2014/1/21 Ildar Karimov notifications@github.com

**Comment by thevery** [comment]
quick ideas about good architecture of the app:
1) download everything via service, not activity
2) download small images for first screen and bigger (and adaptive!) for second
3) reuse/recycle bitmaps
but 2) alone should be enough I think

File: PopularPhotos/res/values/strings.xml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <string name="app_name">Popular photos via Yandex.Fotki</string>
+    <string name="update">Tap to update</string>
+    <string name="startUpdateMsg">Updating started...</string>
```
Comment (thevery) [inline PopularPhotos/res/values/strings.xml:5]: string names should be lowercase_with_underscore

File: PopularPhotos/src/com/gmail/mazinva/PopularPhotos/MainActivity.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+            GridAdapter adapter = new GridAdapter(context, android.R.layout.simple_list_item_1, imagesScaled);
+            gridView.setAdapter(adapter);
+            cursor.close();
+        } catch (MySQLiteException e) {
+            e.printStackTrace();
+        }
+    }
+
+    private void getOrientation() {
```
Comment (thevery) [inline PopularPhotos/src/com/gmail/mazinva/PopularPhotos/MainActivity.java:160]: use layout and layout-land resource folders instead

File: PopularPhotos/src/com/gmail/mazinva/PopularPhotos/SingleImageActivity.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            id++;
+            Cursor cursor = accessor.fetchImageById(id);
+            cursor.moveToFirst();
+
+            byte[] data = cursor.getBlob(cursor.getColumnIndexOrThrow(SQLiteAccessor.KEY_BITMAP_IMAGES));
+            Bitmap bitmap = BitmapFactory.decodeByteArray(data, 0, data.length);
+            ImageView imageView = (ImageView) findViewById(R.id.image);
+            imageView.setImageBitmap(bitmap);
+            cursor.close();
+        } catch (MySQLiteException e) {
```
Comment (thevery) [inline PopularPhotos/src/com/gmail/mazinva/PopularPhotos/SingleImageActivity.java:32]: looks like an ugly hack

**Comment by thevery** [comment]
20-3+1=18 (no major problems in code)

**Pull request closed without merging**